### PR TITLE
xds/cds: add separate fields for cluster name and eds service name

### PR DESCRIPTION
**Pull request by menghanl**
ClusterName and EDSServiceName are two separate fields in the CDS resp.
TD only sets cluster name, not eds service name.

EDS should watch eds service name if it's set, otherwise watch cluster name.
The current eds balancer assumes the name to watch EDS is the cluster name.

LRS needs both, thus when reporting loads, eds balancer sets the cluster name in load reports to the service name.
This PR cleans that up.

### CODE REVIEW

File: xds/internal/client/xds.go [hunk 1]
```diff
@@ -571,40 +571,35 @@ func unmarshalClusterResource(r *anypb.Any, logger *grpclog.PrefixLogger) (strin
 	return cluster.GetName(), cu, nil
 }
 
-func clusterTypeFromCluster(cluster *v3clusterpb.Cluster) (ClusterType, string, []string, error) {
+func clusterTypeFromCluster(cluster *v3clusterpb.Cluster) (ClusterType, string, string, []string, error) {
```
Comment (easwars) [inline xds/internal/client/xds.go:574]: This function seems to be doing a bit too much. It is not only determining the cluster type, but also getting the cluster name, eds service name, list of clusters (for the aggregate case).

Can we simplify things here? Some thoughts that come to mind are:
- Make this function only return the cluster type, and have a switch case in the caller based on the cluster type and set the appropriate fields in the `ClusterUpdate` directly
- Pass the `ClusterUpdate` as a pointer to this function and have it directly set things up. Rename the function appropriately if we are doing that.

File: xds/internal/client/xds.go [hunk 2]
```diff
@@ -571,40 +571,35 @@ func unmarshalClusterResource(r *anypb.Any, logger *grpclog.PrefixLogger) (strin
...
 	if cluster.GetType() == v3clusterpb.Cluster_EDS {
 		if cluster.GetEdsClusterConfig().GetEdsConfig().GetAds() == nil {
-			return 0, "", nil, fmt.Errorf("unexpected edsConfig in response: %+v", cluster)
+			return 0, "", "", nil, fmt.Errorf("unexpected edsConfig in response: %+v", cluster)
```
Comment (easwars) [inline xds/internal/client/xds.go:577]: Can we avoid integer literals here, and use the appropriate const instead.

File: xds/internal/client/client.go [hunk 3]
```diff
@@ -375,9 +375,12 @@ const (
 // interest to the registered CDS watcher.
 type ClusterUpdate struct {
 	ClusterType ClusterType
-	// ServiceName is the service name corresponding to the clusterName which
+	// ClusterName is the service name corresponding to the clusterName which
```
Comment (easwars) [inline xds/internal/client/client.go:378]: This comment needs updating?
Is `ClusterName` the service name?

File: xds/internal/balancer/edsbalancer/config.go [hunk 4]
```diff
@@ -35,8 +35,9 @@ type EDSConfig struct {
 	// FallBackPolicy represents the load balancing config for the
 	// fallback.
 	FallBackPolicy *loadBalancingConfig
-	// Name to use in EDS query.  If not present, defaults to the server
-	// name from the target URI.
+	// Cluster name.
+	ClusterName string
+	// Name to use in EDS query. If not set, use ClusterName.
```
Comment (easwars) [inline xds/internal/balancer/edsbalancer/config.go:40]: Super nit: These comments should also follow the Go documentation style where it starts with the field name.

File: xds/internal/client/xds.go (hunk 1 above)
Comment (menghanl) [reply xds/internal/client/xds.go:574]: I moved code around and deleted this function.
PTAL.

File: xds/internal/client/cds_test.go [hunk 5]
```diff
@@ -139,24 +141,38 @@ func (s) TestValidateCluster_Success(t *testing.T) {
...
 		},
 		{
 			name: "happy-case-aggregate-v3",
 			cluster: &v3clusterpb.Cluster{
+				Name: clusterName,
 				ClusterDiscoveryType: &v3clusterpb.Cluster_ClusterType{
-					ClusterType: &v3clusterpb.Cluster_CustomClusterType{Name: "envoy.clusters.aggregate"},
+					ClusterType: &v3clusterpb.Cluster_CustomClusterType{
+						Name: "envoy.clusters.aggregate",
+						TypedConfig: func() *anypb2.Any {
```
Comment (easwars) [inline xds/internal/client/cds_test.go:157]: We now have a utility function to do this:
https://github.com/grpc/grpc-go/blob/master/internal/testutils/marshal_any.go#L30

File: xds/internal/client/cds_test.go [hunk 6]
```diff
@@ -35,6 +36,7 @@ import (
 	xdsinternal "google.golang.org/grpc/internal/xds"
 	"google.golang.org/grpc/internal/xds/env"
 	"google.golang.org/grpc/xds/internal/version"
+	anypb2 "google.golang.org/protobuf/types/known/anypb"
```
Comment (easwars) [inline xds/internal/client/cds_test.go:39]: I guess this is Intellij doing this funky renaming. We might not need to import this package once we use `testutils.MarshalAny()`.

File: xds/internal/client/xds.go [hunk 7]
```diff
@@ -571,62 +571,54 @@ func unmarshalClusterResource(r *anypb.Any, logger *grpclog.PrefixLogger) (strin
...
+	}
+
+	ret := ClusterUpdate{
+		ClusterName: cluster.GetName(),
+		EnableLRS:   cluster.GetLrsServer().GetSelf() != nil,
+		SecurityCfg: sc,
+		MaxRequests: circuitBreakersFromCluster(cluster),
+	}
+
+	// Validate and set cluster type from the response.
```
Comment (easwars) [inline xds/internal/client/xds.go:596]: Does it make more sense to use a `switch` instead? I believe all these `if`s are mutually exclusive, right?

File: xds/internal/client/cds_test.go (hunk 5 above)
Comment (menghanl) [reply xds/internal/client/cds_test.go]: Done

File: xds/internal/client/cds_test.go (hunk 6 above)
Comment (menghanl) [reply xds/internal/client/cds_test.go]: Yes, it's removed.

File: xds/internal/client/xds.go (hunk 7 above)
Comment (menghanl) [reply xds/internal/client/xds.go:596]: Done.

**Review by easwars — APPROVED** [review]

**Pull request merged**